File: internal/orchestration/engine.go
// Package orchestration runs one roundtable turn across the configured
// providers.
//
// A full turn has three phases:
//
//  1. Independent - every provider answers in parallel, blind to the others.
//  2. Peer review - every provider that succeeded reviews all the answers.
//  3. Chair - the selected provider writes the single final answer.
//
// A provider that fails is isolated: the turn continues with whoever is left.
// If the chair itself fails, another successful provider takes over.
package orchestration

import (
	"context"
	"errors"
	"fmt"
	"strings"
	"sync"

	"github.com/steve1603/AgentConnect/internal/config"
	"github.com/steve1603/AgentConnect/internal/providers"
)

// Turn modes.
const (
	ModeFull   = "full"
	ModePanel  = "panel"
	ModeDirect = "direct"
)

// Modes lists every supported mode, in display order.
var Modes = []string{ModeFull, ModePanel, ModeDirect}

// ErrNoAnswer means no provider produced anything usable for this turn.
var ErrNoAnswer = errors.New("no provider produced an answer")

// Event is one progress update emitted while a turn runs. Only the fields
// relevant to a given event type are populated.
type Event struct {
	Type      string   `json:"type"`
	Phase     string   `json:"phase,omitempty"`
	Status    string   `json:"status,omitempty"`
	Provider  string   `json:"provider,omitempty"`
	Label     string   `json:"label,omitempty"`
	Model     string   `json:"model,omitempty"`
	LatencyMS int64    `json:"latency_ms,omitempty"`
	Attempts  int      `json:"attempts,omitempty"`
	Error     string   `json:"error,omitempty"`
	Reason    string   `json:"reason,omitempty"`
	Providers []string `json:"providers,omitempty"`
	Succeeded []string `json:"succeeded,omitempty"`
	Result    *Outcome `json:"result,omitempty"`
	Message   string   `json:"message,omitempty"`
	MessageID int64    `json:"message_id,omitempty"`
}

// Outcome is everything a completed turn produced, including the full trace.
type Outcome struct {
	Mode           string             `json:"mode"`
	FinalText      string             `json:"final_text"`
	Chair          string             `json:"chair,omitempty"`
	ChairRequested string             `json:"chair_requested,omitempty"`
	ChairFallback  bool               `json:"chair_fallback"`
	Answers        []providers.Result `json:"answers"`
	Reviews        []providers.Result `json:"reviews"`
	ChairResult    *providers.Result  `json:"chair_result,omitempty"`
	Failures       []providers.Result `json:"failures"`
}

// Request describes one turn.
type Request struct {
	Text    string
	History []HistoryMessage
	Mode    string
	Chair   string
}

// Engine runs turns against a fixed set of providers.
type Engine struct {
	providers map[string]providers.Provider
	cfg       *config.Config
}

// New builds an engine over the given providers.
func New(p map[string]providers.Provider, cfg *config.Config) *Engine {
	return &Engine{providers: p, cfg: cfg}
}

// Run executes one turn, passing every progress event to emit. It returns the
// outcome, or an error if the turn could not produce an answer at all.
func (e *Engine) Run(ctx context.Context, req Request, emit func(Event)) (*Outcome, error) {
	if len(e.providers) == 0 {
		return nil, errors.New("no providers are configured; add at least one API key to .env")
	}

	mode := req.Mode
	if mode != ModeFull && mode != ModePanel && mode != ModeDirect {
		mode = ModeFull
	}

	chair := req.Chair
	if _, ok := e.providers[chair]; !ok {
		chair = e.cfg.ResolvedChair()
	}
	if _, ok := e.providers[chair]; !ok {
		chair = e.order()[0]
	}

	history := RenderHistory(req.History, e.cfg.HistoryMessageLimit, e.cfg.HistoryCharLimit)
	maxChars := e.cfg.MaxPromptChars

	if mode == ModeDirect {
		return e.runDirect(ctx, req.Text, history, chair, maxChars, emit)
	}

	// -- Phase 1: independent answers -------------------------------------
	order := e.order()
	emit(Event{Type: "phase", Phase: "independent", Status: "start", Providers: order})

	prompt := BuildIndependentPrompt(req.Text, history, maxChars)
	jobs := make(map[string]job, len(order))
	for _, key := range order {
		jobs[key] = job{system: IndependentSystem, prompt: prompt}
	}
	answers := e.runParallel(ctx, "independent", order, jobs, emit)

	succeeded := successesInOrder(answers)
	emit(Event{
		Type: "phase", Phase: "independent", Status: "done",
		Succeeded: keysOf(succeeded),
	})

	if len(succeeded) == 0 {
		return nil, fmt.Errorf("%w: %s", ErrNoAnswer, describeFailures(answers))
	}

	// -- Phase 2: peer review ---------------------------------------------
	var reviews []providers.Result
	switch {
	case mode == ModeFull && len(succeeded) > 1:
		reviewers := keysOf(succeeded)
		emit(Event{Type: "phase", Phase: "review", Status: "start", Providers: reviewers})

		payload := toAnswers(succeeded)
		reviewJobs := make(map[string]job, len(reviewers))
		for _, key := range reviewers {
			reviewJobs[key] = job{
				system: ReviewSystem,
				prompt: BuildReviewPrompt(req.Text, payload, e.providers[key].Label(), history, maxChars),
			}
		}
		reviews = e.runParallel(ctx, "review", reviewers, reviewJobs, emit)

		emit(Event{
			Type: "phase", Phase: "review", Status: "done",
			Succeeded: keysOf(successesInOrder(reviews)),
		})
	case mode == ModeFull:
		emit(Event{
			Type: "phase", Phase: "review", Status: "skipped",
			Reason: "only one provider produced an answer",
		})
	}

	// -- Phase 3: chair synthesis -----------------------------------------
	chairPrompt := BuildChairPrompt(
		req.Text, toAnswers(succeeded), toAnswers(successesInOrder(reviews)), history, maxChars,
	)

	var (
		chairResult   *providers.Result
		chairUsed     string
		chairAttempts []providers.Result
	)

	for _, candidate := range chairCandidates(chair, succeeded) {
		provider := e.providers[candidate]
		emit(Event{
			Type: "provider", Phase: "chair", Provider: candidate,
			Label: provider.Label(), Model: provider.Model(), Status: "running",
		})

		result := provider.Complete(ctx, ChairSystem, chairPrompt)
		chairAttempts = append(chairAttempts, result)
		emit(providerEvent("chair", result))

		if result.OK {
			chairResult = &result
			chairUsed = candidate
			break
		}
	}

	outcome := &Outcome{
		Mode:           mode,
		ChairRequested: chair,
		Answers:        answers,
		Reviews:        reviews,
		Failures:       failuresOf(answers, reviews, chairAttempts),
	}

	if chairResult == nil {
		// Every chair candidate failed. Return the strongest single answer
		// rather than losing the turn.
		emit(Event{
			Type: "phase", Phase: "chair", Status: "failed",
			Reason: "no provider could synthesise; returning the strongest single answer",
		})
		outcome.FinalText = succeeded[0].Text
		outcome.ChairFallback = true
	} else {
		outcome.FinalText = chairResult.Text
		outcome.Chair = chairUsed
		outcome.ChairResult = chairResult
		outcome.ChairFallback = chairUsed != chair
	}

	emit(Event{Type: "complete", Result: outcome})
	return outcome, nil
}

func (e *Engine) runDirect(
	ctx context.Context, request, history, key string, maxChars int, emit func(Event),
) (*Outcome, error) {
	provider := e.providers[key]
	emit(Event{
		Type: "provider", Phase: "direct", Provider: key,
		Label: provider.Label(), Model: provider.Model(), Status: "running",
	})

	result := provider.Complete(ctx, DirectSystem, BuildDirectPrompt(request, history, maxChars))
	emit(providerEvent("direct", result))

	if !result.OK {
		return nil, fmt.Errorf("%w: %s failed: %s", ErrNoAnswer, result.Label, result.Error)
	}

	outcome := &Outcome{
		Mode:           ModeDirect,
		FinalText:      result.Text,
		Chair:          key,
		ChairRequested: key,
		Answers:        []providers.Result{result},
		Reviews:        []providers.Result{},
		Failures:       []providers.Result{},
	}
	emit(Event{Type: "complete", Result: outcome})
	return outcome, nil
}

type job struct {
	system string
	prompt string
}

// runParallel calls several providers concurrently, emitting an event as each
// one starts and again as each one lands. Results come back in the stable
// order given by keys, not in completion order.
func (e *Engine) runParallel(
	ctx context.Context, phase string, keys []string, jobs map[string]job, emit func(Event),
) []providers.Result {
	for _, key := range keys {
		provider := e.providers[key]
		emit(Event{
			Type: "provider", Phase: phase, Provider: key,
			Label: provider.Label(), Model: provider.Model(), Status: "running",
		})
	}

	type landed struct {
		index  int
		result providers.Result
	}

	landings := make(chan landed, len(keys))
	var wg sync.WaitGroup
	for index, key := range keys {
		wg.Add(1)
		go func(index int, key string) {
			defer wg.Done()
			spec := jobs[key]
			landings <- landed{index: index, result: e.providers[key].Complete(ctx, spec.system, spec.prompt)}
		}(index, key)
	}
	go func() {
		wg.Wait()
		close(landings)
	}()

	results := make([]providers.Result, len(keys))
	for l := range landings {
		results[l.index] = l.result
		emit(providerEvent(phase, l.result))
	}
	return results
}

// order lists the configured providers in canonical display order.
func (e *Engine) order() []string {
	var keys []string
	for _, key := range config.Keys {
		if _, ok := e.providers[key]; ok {
			keys = append(keys, key)
		}
	}
	return keys
}

func providerEvent(phase string, result providers.Result) Event {
	status := "ok"
	if !result.OK {
		status = "error"
	}
	return Event{
		Type:      "provider",
		Phase:     phase,
		Provider:  result.Provider,
		Label:     result.Label,
		Model:     result.Model,
		Status:    status,
		LatencyMS: result.LatencyMS,
		Attempts:  result.Attempts,
		Error:     result.Error,
	}
}

// successesInOrder keeps the successful results in canonical provider order.
func successesInOrder(results []providers.Result) []providers.Result {
	byKey := make(map[string]providers.Result, len(results))
	for _, result := range results {
		if result.OK {
			byKey[result.Provider] = result
		}
	}
	var ordered []providers.Result
	for _, key := range config.Keys {
		if result, ok := byKey[key]; ok {
			ordered = append(ordered, result)
		}
	}
	return ordered
}

func keysOf(results []providers.Result) []string {
	keys := make([]string, 0, len(results))
	for _, result := range results {
		keys = append(keys, result.Provider)
	}
	return keys
}

func toAnswers(results []providers.Result) []Answer {
	answers := make([]Answer, 0, len(results))
	for _, result := range results {
		answers = append(answers, Answer{Label: result.Label, Model: result.Model, Text: result.Text})
	}
	return answers
}

func failuresOf(groups ...[]providers.Result) []providers.Result {
	failures := []providers.Result{}
	for _, group := range groups {
		for _, result := range group {
			if !result.OK {
				failures = append(failures, result)
			}
		}
	}
	return failures
}

// chairCandidates is the chair to try first, then the fallbacks. A requested
// chair that already failed this turn is demoted behind the providers that
// answered, but is still tried, since its failure may be transient.
func chairCandidates(requested string, succeeded []providers.Result) []string {
	var others []string
	requestedSucceeded := false
	for _, result := range succeeded {
		if result.Provider == requested {
			requestedSucceeded = true
			continue
		}
		others = append(others, result.Provider)
	}
	if requestedSucceeded {
		return append([]string{requested}, others...)
	}
	return append(others, requested)
}

func describeFailures(results []providers.Result) string {
	parts := make([]string, 0, len(results))
	for _, result := range results {
		if !result.OK {
			parts = append(parts, result.Label+": "+result.Error)
		}
	}
	return strings.Join(parts, "; ")
}

File: app/orchestration/engine.py
"""The roundtable orchestrator.

A full turn runs three phases:

1. Independent - every configured provider answers the request in parallel,
   without seeing any other answer.
2. Peer review - every provider that succeeded in phase 1 receives all
   successful answers and critiques them, again in parallel.
3. Chair - the selected provider receives the request, the answers, and the
   reviews, and writes the single final answer.

A provider that fails is isolated: the turn continues with whoever is left.
If the chair itself fails, another successful provider takes over.
"""

from __future__ import annotations

import asyncio
import logging
from dataclasses import dataclass, field
from typing import Any, AsyncIterator, Mapping, Sequence

from ..config import PROVIDER_KEYS, Settings
from ..providers.base import Provider, ProviderResult
from . import prompts

logger = logging.getLogger(__name__)

MODE_FULL = "full"
MODE_PANEL = "panel"
MODE_DIRECT = "direct"
MODES = (MODE_FULL, MODE_PANEL, MODE_DIRECT)


class OrchestrationError(RuntimeError):
    """No provider was able to produce an answer for this turn."""


@dataclass(slots=True)
class TurnOutcome:
    """Everything a completed turn produced, including the full trace."""

    mode: str
    final_text: str
    chair: str | None = None
    chair_requested: str | None = None
    chair_fallback: bool = False
    answers: list[dict[str, Any]] = field(default_factory=list)
    reviews: list[dict[str, Any]] = field(default_factory=list)
    chair_result: dict[str, Any] | None = None
    failures: list[dict[str, Any]] = field(default_factory=list)

    def to_dict(self) -> dict[str, Any]:
        return {
            "mode": self.mode,
            "final_text": self.final_text,
            "chair": self.chair,
            "chair_requested": self.chair_requested,
            "chair_fallback": self.chair_fallback,
            "answers": self.answers,
            "reviews": self.reviews,
            "chair_result": self.chair_result,
            "failures": self.failures,
        }


class RoundtableEngine:
    """Runs one turn across the configured providers."""

    def __init__(self, providers: Mapping[str, Provider], settings: Settings) -> None:
        self.providers = dict(providers)
        self.settings = settings

    # -- public API --------------------------------------------------------

    async def run(
        self,
        request: str,
        history: Sequence[Mapping[str, str]] = (),
        *,
        mode: str = MODE_FULL,
        chair: str | None = None,
    ) -> AsyncIterator[dict[str, Any]]:
        """Run a turn, yielding progress events and finally a `complete` event."""
        if mode not in MODES:
            mode = MODE_FULL
        if not self.providers:
            raise OrchestrationError(
                "No providers are configured. Add at least one API key to .env."
            )

        requested_chair = chair if chair in self.providers else self.settings.resolved_default_chair()
        if requested_chair not in self.providers:
            requested_chair = next(iter(self.providers))

        rendered_history = prompts.render_history(
            history,
            message_limit=self.settings.history_message_limit,
            char_limit=self.settings.history_char_limit,
        )
        max_chars = self.settings.max_prompt_chars

        if mode == MODE_DIRECT:
            async for event in self._run_direct(
                request, rendered_history, requested_chair, max_chars
            ):
                yield event
            return

        # -- Phase 1: independent answers ----------------------------------
        order = [k for k in PROVIDER_KEYS if k in self.providers]
        yield {"type": "phase", "phase": "independent", "status": "start", "providers": order}

        independent_prompt = prompts.build_independent_prompt(
            request, rendered_history, max_chars=max_chars
        )
        answers: list[ProviderResult] = []
        async for event, result in self._run_parallel(
            "independent",
            {key: (prompts.INDEPENDENT_SYSTEM, independent_prompt) for key in order},
        ):
            if event is not None:
                yield event
            if result is not None:
                answers.append(result)

        successful = _ordered_successes(answers)
        yield {
            "type": "phase",
            "phase": "independent",
            "status": "done",
            "succeeded": [r.provider for r in successful],
        }

        if not successful:
            raise OrchestrationError(
                "Every provider failed on the first round: "
                + "; ".join(f"{r.label}: {r.error}" for r in answers)
            )

        # -- Phase 2: peer review ------------------------------------------
        reviews: list[ProviderResult] = []
        if mode == MODE_FULL and len(successful) > 1:
            reviewer_keys = [r.provider for r in successful]
            yield {"type": "phase", "phase": "review", "status": "start", "providers": reviewer_keys}

            answer_payload = [_answer_payload(r) for r in successful]
            review_jobs = {
                key: (
                    prompts.REVIEW_SYSTEM,
                    prompts.build_review_prompt(
                        request,
                        answer_payload,
                        self.providers[key].label,
                        rendered_history,
                        max_chars=max_chars,
                    ),
                )
                for key in reviewer_keys
            }
            async for event, result in self._run_parallel("review", review_jobs):
                if event is not None:
                    yield event
                if result is not None:
                    reviews.append(result)

            yield {
                "type": "phase",
                "phase": "review",
                "status": "done",
                "succeeded": [r.provider for r in reviews if r.ok],
            }
        elif mode == MODE_FULL:
            yield {
                "type": "phase",
                "phase": "review",
                "status": "skipped",
                "reason": "only one provider produced an answer",
            }

        # -- Phase 3: chair synthesis --------------------------------------
        chair_prompt = prompts.build_chair_prompt(
            request,
            [_answer_payload(r) for r in successful],
            [_answer_payload(r) for r in reviews if r.ok],
            rendered_history,
            max_chars=max_chars,
        )

        chair_result: ProviderResult | None = None
        chair_used: str | None = None
        chair_attempts: list[ProviderResult] = []

        for candidate in _chair_candidates(requested_chair, successful):
            yield {
                "type": "provider",
                "phase": "chair",
                "provider": candidate,
                "status": "running",
                "model": self.providers[candidate].model,
            }
            result = await self.providers[candidate].complete(prompts.CHAIR_SYSTEM, chair_prompt)
            chair_attempts.append(result)
            yield _provider_event("chair", result)
            if result.ok:
                chair_result = result
                chair_used = candidate
                break

        failures = [r.to_dict() for r in answers + reviews + chair_attempts if not r.ok]

        if chair_result is None:
            # Every chair candidate failed; fall back to the best raw answer
            # rather than losing the whole turn.
            fallback = successful[0]
            yield {
                "type": "phase",
                "phase": "chair",
                "status": "failed",
                "reason": "no provider could synthesise; returning the strongest single answer",
            }
            outcome = TurnOutcome(
                mode=mode,
                final_text=fallback.text,
                chair=None,
                chair_requested=requested_chair,
                chair_fallback=True,
                answers=[r.to_dict() for r in answers],
                reviews=[r.to_dict() for r in reviews],
                chair_result=None,
                failures=failures,
            )
        else:
            outcome = TurnOutcome(
                mode=mode,
                final_text=chair_result.text,
                chair=chair_used,
                chair_requested=requested_chair,
                chair_fallback=chair_used != requested_chair,
                answers=[r.to_dict() for r in answers],
                reviews=[r.to_dict() for r in reviews],
                chair_result=chair_result.to_dict(),
                failures=failures,
            )

        yield {"type": "complete", "result": outcome.to_dict()}

    # -- internals ---------------------------------------------------------

    async def _run_direct(
        self,
        request: str,
        history: str,
        provider_key: str,
        max_chars: int,
    ) -> AsyncIterator[dict[str, Any]]:
        provider = self.providers[provider_key]
        yield {
            "type": "provider",
            "phase": "direct",
            "provider": provider_key,
            "status": "running",
            "model": provider.model,
        }
        result = await provider.complete(
            prompts.DIRECT_SYSTEM,
            prompts.build_direct_prompt(request, history, max_chars=max_chars),
        )
        yield _provider_event("direct", result)

        if not result.ok:
            raise OrchestrationError(f"{result.label} failed: {result.error}")

        outcome = TurnOutcome(
            mode=MODE_DIRECT,
            final_text=result.text,
            chair=provider_key,
            chair_requested=provider_key,
            answers=[result.to_dict()],
        )
        yield {"type": "complete", "result": outcome.to_dict()}

    async def _run_parallel(
        self,
        phase: str,
        jobs: Mapping[str, tuple[str, str]],
    ) -> AsyncIterator[tuple[dict[str, Any] | None, ProviderResult | None]]:
        """Run several providers concurrently, streaming events as they land."""
        queue: asyncio.Queue[ProviderResult] = asyncio.Queue()

        async def worker(key: str, system: str, prompt: str) -> None:
            result = await self.providers[key].complete(system, prompt)
            await queue.put(result)

        tasks = [
            asyncio.create_task(worker(key, system, prompt))
            for key, (system, prompt) in jobs.items()
        ]

        for key in jobs:
            yield (
                {
                    "type": "provider",
                    "phase": phase,
                    "provider": key,
                    "status": "running",
                    "model": self.providers[key].model,
                },
                None,
            )

        try:
            for _ in range(len(tasks)):
                result = await queue.get()
                yield (_provider_event(phase, result), result)
        finally:
            for task in tasks:
                if not task.done():
                    task.cancel()
            await asyncio.gather(*tasks, return_exceptions=True)


def _provider_event(phase: str, result: ProviderResult) -> dict[str, Any]:
    return {
        "type": "provider",
        "phase": phase,
        "provider": result.provider,
        "label": result.label,
        "model": result.model,
        "status": "ok" if result.ok else "error",
        "latency_ms": result.latency_ms,
        "attempts": result.attempts,
        "error": result.error,
    }


def _answer_payload(result: ProviderResult) -> dict[str, str]:
    return {"label": result.label, "model": result.model, "text": result.text}


def _ordered_successes(results: Sequence[ProviderResult]) -> list[ProviderResult]:
    """Successful results in stable provider order, not completion order."""
    by_key = {r.provider: r for r in results if r.ok}
    return [by_key[k] for k in PROVIDER_KEYS if k in by_key]


def _chair_candidates(
    requested: str,
    successful: Sequence[ProviderResult],
) -> list[str]:
    """The chair to try first, then the fallbacks, in a stable order.

    A requested chair that already failed this turn is demoted behind the
    providers that answered successfully, so one dead vendor does not stall
    synthesis - but it is still tried, since its failure may be transient.
    """
    others = [r.provider for r in successful if r.provider != requested]
    if any(r.provider == requested for r in successful):
        return [requested, *others]
    return [*others, requested]
